### ISSUE TITLE
Use the Event.RENDER event to interact with the flash frame skipping.

### ISSUE THREAD

**Issue by henke37** [Gamua/Starling-Framework#655]
The Flash player automatically skips rendering frames when it is too far behind. Starling does not, instead it tells all parts about how long it is since the last update.

Starling should obviously continue with the "how long since last update" feature, if only for compatibility reasons.

But starling should not be rendering frames when it is way behind. I recommend listening for the Event.RENDER event to let the Flash player itself decide when to render new frames.

**Comment by PrimaryFeather** [comment]
I had a look at it, and it's unfortunately not as easy as I thought.

Using the "RENDER" event alone is not an option, because a game's logic might need some code to be executed even though it's not being displayed (think of a multiplayer game that is minimized).

So I'd need to split up rendering and "advanceTime"; as Josh suggested in this [forum thread](http://forum.starling-framework.org/topic/what-event-to-render-on#post-41895), I'd ideally call "nativeStage.invalidate" in the ENTER_FRAME or EXIT_FRAME handler and then do the rendering in the RENDER event.

However, Starling currently contains a public "nextFrame()" method which calls both "advanceTime" and "render". This method is used by people who use a shared render context, e.g. when you combine Starling and Away3D. I'd have to remove the "render" call from that method, but that would break all those games.

It might be a change that can be communicated to developers; but the question is if it's worth causing developers these troubles — as I see it, there are not that many issues with the current way of doing it. So I'm tending to delay this change until Starling 2.0.

Any other thoughts? Is this a critical issue for many developers? Thanks in advance for any comments!

**Comment by henke37** [comment]
Since it is a small change for a few affected developers, I think it is simple.

If they are calling nextFrame() manually then it's their burden of dealing with the correct timing. A reasonable setup can be provided for legacy purposes. That is, leave the method as is. People who want the benefits from decoupling game logic and rendering will have to do the change themselves.

Meanwhile, developers who only work with Starling will benefit from this silently.

**Comment by PrimaryFeather** [comment]
Hm, you're right — I could simply deprecate "nextFrame", that's not a bad idea!

Still, could you elaborate a little more about the exact problems that arise with the current way of doing it? Since all Starling animations are time-based (and thus should not be affected), I guess there are problems when combining Starling and the classic display list?

**Comment by henke37** [comment]
I am not sure about the exact details, especially not when it comes to stage3d interaction, but here is what I know:

The Render event happens after the interactive wait period. The enter frame happens before. Not listening for the render event means that output may be stale if the state is updated by an event listener in the time window between the two events.

Additionally, the render event is designed as a way to avoid redundant work. For example, the flash components defer the layout logic until the render event, this allows the dependent properties to be changed as much as the user wants without rerunning the layout logic and throwing away all the results but the last one.

The Render event happens if there is an extra rendering requested in response to an interactive event. This allows for "extra smooth" animation in response to events, like mouse cursors and text typing.

The render event does not happen if the flash player is in frame skipping mode and the current frame is being skipped.

Basically, frame skipping works by omitting the expensive rendering step and only doing the hopefully cheap state updating of the enter frame event. Ideally this allows the player to catch up if it is falling behind and to decouple the state update rate from the effective rendering rate.

**Comment by PrimaryFeather** [comment]
Thanks for the detailed description, @henke37! I'm roughly familiar with the logic behind that, but what I meant was actually: what are the main problems you have had (or heard from others) because of Starling doing it's rendering in the `ENTER_FRAME` event? Or in other words: what immediate issues would be solved for you if I'm doing rendering only in response to a native `RENDER` event?

Because if I simply change the rendering to happen in the native `RENDER` event, that's only a part of the equation; Starling e.g. does not have it's own version of `RENDER`, so people could e.g. not defer layout logic until before rendering. Thus, there's also no `updateAfterEvent()` method on the event right now (because that doesn't make sense with the current code).

And "catching up if it's falling behind" should not be an issue in Starling, because all animations are time-based. So if you'll never fall behind; instead, the `passedTime` will be higher. Except, of course, if you're using frame based animation (that's why I mentioned the combination of Starling and the classic display list).

So I'm asking myself if making this change will help much, or if we'd have to scrutinize the complete rendering process to enhance the situation (maybe add a Starling-`RENDER` event, etc).

And, by the way, to anyone reading this: any other voices, suggestions and feedback is welcome! :-)

**Comment by henke37** [comment]
Obviously any attempts at synchronizing with frame based animations will behave poorly. And the constrained profile is known to have "interesting ideas" about how to link Stage3D and the normal rendering.

My personal setup is to let starling do just the rendering. All my animation and movement is done with a native enter frame listener.

The main concern I can find is the wasted rendering when the flash player is in frame skipping mode. My interpretation of the scenario is as follows.

It is the start of a new frame. The player decides that it is way behind for some reason. It dispatches the enter frame event that starling is listening for. Starling detects that it's been a fair amount of realtime since the last event and updates state a bit further than normal. It then renders and presents. So far, so good. The flash player then enters the interactive wait period, only to immediately leave it, since it has no time to wait. The player then ends the frame, skipping native rendering. Still no problem yet.

Now we start a second frame. The player is still behind. It dutifully dispatches the enter frame event and starling is still listening. Starling will now detect that little to no realtime has elapsed and updates with a corresponding minimal length. Starling then renders and presents again. The interactive wait period is cut short again. The flash player ends the frame, again skipping native rendering. This happens for a total of zero or more times.

Finally we start one more frame. The player decides that it has finally caught up. It dispatches the event again. Starling detects one last time that little to no time has elapsed. Starling renders and presents again. The flash player enters the interactive wait period and waits this time. Once the wait is over, the player ends the frame. The player now finally renders the native display tree.

Two observations can be made.

The first observation is that all work that starling does will add to the load and push the flash player further and further from catching up to where it wants to be. If the rendering and presenting blocks for a major amount of time, as presenting might do, then the player might never be able to catch up.

The second observation is that starling is wasting time rerendering a scene that has no meaningful changes. This is a waste of resources. This extra load could in theory be affecting the effective framerate of the starling scene too.

**Comment by neolit123** [comment]
@henke37,

by frame skipping in the AVM2, do you refer to the frame rate reduction that happens when the race track is adjusted to accommodate the longer slice lengths due to higher code and render times?
i ask because to my knowledge frames are never skipped, but lenghtened.

anyhow, i think a test project with some heavy code shows some problems with the Event.RENDER concept (at least to me that is).

big question:
where and when to call Stage.invalidate()?

for example, and from my testes, if you just call Stage.invalidate() on EXIT_FRAME that will always trigger Event.RENDER, no matter if the projector is minimized and no matter how much the frame rate has dropped due to heavy code.

so some estimations when to call Stage.invalidate() can be made:
ENTER_FRAME can be used to get the length of the previous frame, while EXIT_FRAME can be used to get the code execution time of the current or previous frame, and also the render time of the previous frame can be obtained by subtraction.

but Stage.invalidate() is a method that matters for the current frame and there is a bit of a time domain discretization problem here. i can get this to skip render calls if i say "ok, last frame took way too long, there is probably no time to render this one", however that will be just an estimate, because we have no way of knowing how much the render time of the current frame will be, unless we have a running average or something - still an estimate though.

perhaps you can provide such a minimal test project with your idea and correct implementation of the Event.RENDER callback?

if this becomes part of Starling eventually, i suggest an optional flag mUseRenderEvent, which if false  - runs in compatibility mode (using ENTER_FRAME as it currently is).

**Comment by henke37** [comment]
You are wrong. The render event does not happen for every exit frame, only the ones the player decides to render. Your observation that it is not omitted when the player is not being visible is likely true.

It takes a major load to trigger frame skipping. Try working with a project that has an overload of native vector content. I have [an example](http://megaswf.com/serve/1389804/) of the render event being used to check for frameskipping.

**Comment by neolit123** [comment]
that's odd,

i've downloaded the PerformanceMonitor class to test with it and i still can't get it to "skip frames".
```
package
{
	import flash.display.Sprite;
	import flash.events.Event;

	public class Main extends Sprite
	{
		// case (1) or (2) doesn't matter

		public function Main():void
		{
			const pm:PerformanceMonitor = new PerformanceMonitor();
			addChild(pm);

			addEventListener(Event.ENTER_FRAME, enterFrame);
			// (1)
			// addEventListener(Event.RENDER, render);
		}

		public function enterFrame(_e:Event):void
		{
			// (1)
			// stage.invalidate();

			// (2)
			heavyCode();
		}

		public function render(_e:Event):void
		{
			heavyCode();
		}

		private var entities:uint = 0;

		private function heavyCode():void
		{
			const MAX_ENTITIES:uint = 2000;
			const ENTITIES_PER_FRAME:uint = 100;

			for (var i:int = 0; i < ENTITIES_PER_FRAME; i++) {
				const dim:uint = 100;
				graphics.beginFill(Math.random() * 0xffffff, 0.5);
				graphics.drawRect(Math.random() * stage.stageWidth, Math.random() * stage.stageHeight, dim, dim);
				graphics.endFill();
				entities++;
				if (entities > MAX_ENTITIES) {
					entities = 0;
					graphics.clear();
				}
			}
		}
	}
}
```

the PerformanceMonitor rectangle stays gray and some of the flags are:
realtime = true, totalSkippedFrames = 0 at all times.
tested inside Chrome with HTML, the native SWF projector and latest Adobe AIR with two different compilers.

i would consider the above heavyCode() quite vector heavy as it drops down from 60fps to 1fps progressively as more transparent rectangles are added.

perhaps my test case is incorrect?

**Comment by PrimaryFeather** [comment]
Thanks a lot for taking the time to experiment with this, Lubomir! (And sorry for my late reply.)
I just wanted to try out those tests myself, but I couldn't find that PerformanceMonitor class anywhere. Could you send me that link?

**Comment by henke37** [comment]
The class is right here. https://www.assembla.com/code/HTools/subversion/nodes/238/HTools/FPS/PerformanceMonitor.as

Quite trivial really, all it does is counting how many times an enter frame event has happened without a render event beforehand.

As for the test I linked above, try running it at a high (modern fullscreen) resolution.

**Comment by neolit123** [comment]
@PrimaryFeather: no problem at all. something strange is happening as i clearly see @henke37's intro/demo animation, having the PerformanceMonitor class blinking orange and red if frames are supposedly "skipped", but for my simple test above which is actually quite CPU heavy the same doesn't happen.

this could be something that has changed between SWF versions. but either way, i question the feature request unless you can reproduce it with Starling in a consistent manner.

**Comment by PrimaryFeather** [comment]
I just tried that sample as well, and I had almost the same results as you had, @neolit123. I tried several different browsers on both Windows and Mac, and none showed any frame skipping, no matter what I did. The only combination that skipped any frames was Chrome on OS X with PPAPI; and even then it was not more than one frame in about 5 seconds.

I didn't try any AIR app yet, though. In any case, I totally agree: before implementing this, we need to understand if and when frame skipping is happening.

**Comment by PrimaryFeather** [comment]
@henke37, the sample that you linked — is that classic display list content, or Stage3D?

**Comment by henke37** [comment]
My example is classic content.

**Comment by henke37** [comment]
I have just noticed that this design detail can have a connection to one of the flash and starling renderers lagging behind the other. Like with issue # 543.
